Allow to combine read-only and read-write disks - COW policy

azat opened the pull request:
It was possible before, but it was not deterministic, since new data could be inserted into the read-only disk and this will fail.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to combine read-only and read-write disks in storage policy (as multiple volumes, or multiple disks). This allows to read data from the entire volume, while inserts will prefer the writable disk (i.e. Copy-on-Write storage policy)

azat committed:
Allow to combine read-only and read-write disks

It was possible before, but it was not deterministic, since new data
could be inserted into the read-only disk and this will fail.

vdimir reviewed (approved):
Changelog entry could be a bit more verbose, we can add something like "Allow to combine read-only and read-write disks in a single volume in configuration. In this setup, data can be read from the entire volume, while inserts always prefer the writable disk"

azat commented:
@vdimir thanks! Updated.

azat committed:
Do not try to reserve space for cloning to the same disk

CI: https://s3.amazonaws.com/clickhouse-test-reports/75862/8ed6f7d26da34dfd92724c72199fbe8b6fbf7814/integration_tests__aarch64__2_4_.html

azat committed:
tests: cover both multi-volume and disk in test_cow_policy

azat committed:
Avoid hanging during DROP of tables with read-only and read-write disks

Note that for parts removal it is OK that isStoredOnReadonlyDisk()
includes isWriteOnce() check, since the removal will be done via hard
links which are not supported

As for regular disks, there is enough isReadOnly() check.

azat committed:
Add logging while skipping cleanup due to read-only/broken/write-once disk in MergeTree

azat commented:
>[Stateless tests (ubsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=75862&sha=32549c936c420ba1345e84d89ffebf08135247b5&name_0=PR&name_1=Stateless%20tests%20%28ubsan%29) — fail: 1, passed: 7533, skipped: 109

- https://github.com/ClickHouse/ClickHouse/issues/75966
